accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,871 @@
+            value `iterator`. Otherwise, evaluate `spread.iterator` to
+            `iterator`. *This will deliberately throw an exception if `spread`
+            is `null` and `element` is not null-aware.*
+
+        1.  Loop:
+
+            1.  If `iterator.moveNext()` returns `false`, exit the loop.
+
+            1.  Evaluate `iterator.current` and append it to `result`. *This
+                will be a MapEntry in a map literal, or any object for a list or
+                set literal.*
+
+        The `iterator` API may not be the most efficient way to traverse the
+        items in a collection. In order to give implementations more room to
+        optimize, we loosen the semantics:
+
+        *   If `spread` is an object whose class implements List, Queue, or Set
+            (all from `dart:core`), an implementation *may* choose to call
+            `length` on the object. This may let it allocate space for the
+            resulting collection more efficiently. Classes that implement List
+            are expected to have an efficient, side-effect free implementation
+            of `length`.
+
+        *   If `spread` is an object whose class implements List from
+            `dart:core`, an implementation may choose to call `[]` to access
+            elements from the list. If it does so, it will only pass indexes `>=
+            0` and `<` the value returned by `length`.
+
+        A Dart implementation may detect whether these options apply at compile
+        time based on the static type of `spread` or at runtime based on the
+        actual value.
+
+1.  Else, if `element` is an `ifElement`:
+
+    1.  Evaluate the condition expression to a value `condition`.
+
+    1.  If the Boolean conversion of `condition` is `true`:
+
+        1.  Evaluate the "then" element using this procedure.
+
+    1.  Else, if there is an "else" element of the `if`:
+
+        1.  Evaluate the "else" element using this procedure.
+
+1.  Else, if `element` is a synchronous `forElement` for a `for-in` loop:
+
+    1.  Evaluate the iterator expression to a value `sequence`.
+
+    1.  Evaluate `sequence.iterator` to a value `iterator`.
+
+    1.  Loop:
+
+        1.  If the Boolean conversion of `iterator.moveNext()` does not return
+            `true`, exit the loop.
+
+        1.  If the `for-in` element declares a variable, create a new namespace
+            and a fresh `variable` for it. Otherwise, use the existing
+            `variable` it refers to.
+
+        1.  Bind `variable` to the result of evaluating `iterator.current`.
+
+        1.  Evaluate the body element using this procedure in the scope of
+            `variable`.
+
+1.  Else, if `element` is an asynchronous `await for-in` element:
+
+    1.  Evaluate the stream expression to a value `stream`. It is a dynamic
+        error if `stream` is not an instance of a class that implements
+        `Stream`.
+
```

Comment:
1. Pause the subscription of any surrounding `await for` loop in the current method.
1. Let `streamDone` be a value implementing `Future<Null>`.